org.rdkit.knime.nodes/src/org/rdkit/knime/nodes/twocomponentreaction2/RDKitTwoComponentReactionNodeModel.java

```diff
@@ -339,214 +342,173 @@ protected DataTableSpec getOutputTableSpec(final int outPort,
+				 *         equal to EMPTY_RESULT_DUE_TO_LACK_OF_ROWS processing will be stopped
+				 *         (successfully). If the result is null, we had a missing reactant 1.
 				 */
 				@Override
-				protected List<DataRow> compute(final DataRow row, final long index) throws Exception {
+				protected List<DataRow> compute(final Pair<Pair<DataRow, Long>, Pair<DataRow, Long>> reactantPair, final long reactionIndex)
+						throws Exception {
+					DataRow row = reactantPair.getFirst().getFirst();
+					long index = reactantPair.getFirst().getSecond();
+					DataRow otherReactant = reactantPair.getSecond().getFirst();
+					long index2 = reactantPair.getSecond().getSecond();
+					
 					final long uniqueWaveId = createUniqueCleanupWaveId();
 					List<DataRow> listNewRows = null;
 
 					boolean bFoundIncluded = false;
 					ROMol mol1 = null;
 
 					try {
-						final RandomAccessRowIterator rowAccess = rowAccessReactant2.get();
-
-						if (rowAccess != null) {
-							// Iterate through all second reactant rows for each first reactant
-							if (bMatrixExpansion) {
-								rowAccess.resetIterator();
-								final long subUniqueWaveId = createUniqueCleanupWaveId();
-								int index2 = 0;
-
-								try {
-									while (rowAccess.hasNext()) {
-										if (isReactionIncluded(index, index2)) {
-											bFoundIncluded = true;
-											if (mol1 == null) {
-												mol1 = markForCleanup(arrInputDataInfo[0][INPUT_COLUMN_REACTANT1].getROMol(row), uniqueWaveId);
-												if (mol1 == null) {
-													// Nothing to calculate, if cell is empty
-													break;
-												}
-											}
+						if (bMatrixExpansion) {
+							// there's one subUniqueWaveId per first reactant, so we use index as a key into a Map
+							final long subUniqueWaveId = uniqueWaveId;	
+
+							if (isReactionIncluded(index, index2)) {
+								bFoundIncluded = true;
+								mol1 = markForCleanup(arrInputDataInfo[0][INPUT_COLUMN_REACTANT1].getROMol(row),
+										uniqueWaveId);
+								if (mol1 != null) { // Nothing to calculate, if cell is empty
+									// Additional data cells
+									final List<DataCell> listAdditionalCells1;
+									if (m_modelAdditionalColumnsEnabled.getBooleanValue()) {
+										listAdditionalCells1 = new ArrayList<>();
+										for (int iReactant1AdditionalColumnIndex : listReactant1AdditionalColumnIndexes) {
+											listAdditionalCells1
+													.add(row.getCell(iReactant1AdditionalColumnIndex));
+										}
+									} else {
+										listAdditionalCells1 = null;
+									}
+
+									listNewRows = processWithSecondReactant(mol1, listAdditionalCells1,
+											otherReactant, listNewRows, subUniqueWaveId, (int) index,
+											(int) index2);
+								}
+							} 
+						}
+
+						// Or: Just take the second reactant with the same row index
+						else {
+							if (index == index2) {
+								if (isReactionIncluded(index, index)) {

```

Comment:
This could be simplified to
```java
						else if (index == index2 && isReactionIncluded(index, index)) {
```
which would allow to remove two sets of curly brackets and make the code more readable.